hudi-utilities/src/main/java/org/apache/hudi/utilities/HoodieClusteringJob.java

```diff
@@ -209,8 +209,7 @@ private int doCluster(JavaSparkContext jsc) throws Exception {
         // Instant time is not specified
         // Find the earliest scheduled clustering instant for execution
         Option<HoodieInstant> firstClusteringInstant =
-            metaClient.getActiveTimeline().firstInstant(
-                HoodieTimeline.REPLACE_COMMIT_ACTION, HoodieInstant.State.REQUESTED);
+            metaClient.getActiveTimeline().filterPendingReplaceTimeline().firstInstant();
```

Comment:
But how could we know that the instant in pending state is sill executing or already finished? There could be multiple offline compaction/clustering jobs running concurrently.

Reply:
You are right. This change applies only to serial execution. It will has some problems for concurrent execution. I try to adjust it. thanks

Reply:
Yeah, can you explain the new strategy literally a little?

Reply:
This is intentional because we should not execute a clustering instant which is already inflight.  If a replacecommit is inflight and the job failed, the right process is to roll back the inflight clustering to requested state first, see: https://github.com/apache/hudi/blob/master/hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/client/SparkRDDTableServiceClient.java#L198

Reply:
> This is intentional because we should not execute a clustering instant which is already inflight. If a replacecommit is inflight and the job failed, the right process is to roll back the inflight clustering to requested state first, see: https://github.com/apache/hudi/blob/master/hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/client/SparkRDDTableServiceClient.java#L198

We should not execute a clustering instant which is already inflight is correct. But if a replacecommit is inflight and the job failed, variable `firstClusteringInstant` is empty for the failed clustering instant in next clustering job. It will throw an exception `throw new HoodieClusteringException("There is no scheduled clustering in the table.");`  This will cause the rollback method not to be called in `SparkRDDTableServiceClient#cluster` code.
```
  Option<HoodieInstant> firstClusteringInstant =
      metaClient.getActiveTimeline().firstInstant(
                HoodieTimeline.REPLACE_COMMIT_ACTION, HoodieInstant.State.REQUESTED);
  ...
  // will not call cluster method for the failed clustering instant
  Option<HoodieCommitMetadata> commitMetadata = client.cluster(cfg.clusteringInstantTime).getCommitMetadata();
````

As **danny0405** said, it is not applicable if multiple offline compaction/clustering jobs running concurrently. I will try to figure out how to adjust.

Reply:
One idea is to align the cluster job with the write client behavior to first roll back the inflight clustering.

Reply:
+1

Reply:
We already support align the clustering job with the write client behavior. But when the write traffic is too large, we need to start an independent job to schedule the compaction or clustering plan.

Reply:
The `writeClient.cluster` internally rolls back the inflight clustering (replacecommit) if exists, so we should be good here.
```
public HoodieWriteMetadata<O> cluster(String clusteringInstant, boolean shouldComplete) {
    HoodieTable<?, I, ?, T> table = createTable(config, context.getHadoopConf().get());
    HoodieTimeline pendingClusteringTimeline = table.getActiveTimeline().filterPendingReplaceTimeline();
    HoodieInstant inflightInstant = HoodieTimeline.getReplaceCommitInflightInstant(clusteringInstant);
    if (pendingClusteringTimeline.containsInstant(inflightInstant)) {
      table.rollbackInflightClustering(inflightInstant, commitToRollback -> getPendingRollbackInfo(table.getMetaClient(), commitToRollback, false));
      table.getMetaClient().reloadActiveTimeline();
    }
    clusteringTimer = metrics.getClusteringCtx();
    LOG.info("Starting clustering at " + clusteringInstant);
...
```